File: python/src/coding_interview/presentation/portfolio_controller.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal, InvalidOperation

from coding_interview.application.usecase.exceptions import InvalidPortfolioError
from coding_interview.application.usecase.portfolio.get_latest_portfolio_usecase import GetLatestPortfolioUsecase
from coding_interview.application.usecase.portfolio.update_portfolio_usecase import (
    UpdatePortfolioItemInput,
    UpdatePortfolioUsecase,
    UpdatePortfolioUsecaseInput,
)
from coding_interview.domain.stock_symbol import StockSymbol
from coding_interview.presentation.exceptions import BadRequestException


@dataclass(frozen=True)
class PortfolioItemDto:
    symbol: str
    rate: str


@dataclass(frozen=True)
class GetOptimalPortfolioResponse:
    portfolios: tuple[PortfolioItemDto, ...]


@dataclass(frozen=True)
class UpdateOptimalPortfolioRequest:
    portfolios: tuple[PortfolioItemDto, ...]


class PortfolioController:
    def __init__(
        self,
        get_latest_portfolio_usecase: GetLatestPortfolioUsecase,
        update_portfolio_usecase: UpdatePortfolioUsecase,
    ) -> None:
        self._get_latest_portfolio_usecase = get_latest_portfolio_usecase
        self._update_portfolio_usecase = update_portfolio_usecase

    def get_optimal_portfolio(self) -> GetOptimalPortfolioResponse:
        out = self._get_latest_portfolio_usecase.run()
        return GetOptimalPortfolioResponse(
            portfolios=tuple(PortfolioItemDto(symbol=str(i.symbol), rate=str(i.rate)) for i in out.items)
        )

    def update_optimal_portfolio(self, req: UpdateOptimalPortfolioRequest) -> None:
        items: list[UpdatePortfolioItemInput] = []
        for dto in req.portfolios:
            sym = StockSymbol.from_string(dto.symbol)
            if sym is None:
                raise BadRequestException(f"unknown symbol: {dto.symbol}")
            try:
                rate = Decimal(dto.rate)
            except (InvalidOperation, Exception):
                raise BadRequestException(f"invalid rate: {dto.rate}")
            items.append(UpdatePortfolioItemInput(sym, rate))
        try:
            self._update_portfolio_usecase.run(UpdatePortfolioUsecaseInput(tuple(items)))
        except InvalidPortfolioError as e:
            raise BadRequestException(e.reason)

File: ruby/lib/coding_interview/presentation/portfolio_controller.rb
require "bigdecimal"
require_relative "presentation_exception"
require_relative "../domain/stock_symbol"
require_relative "../application/usecase/portfolio/get_latest_portfolio_usecase"
require_relative "../application/usecase/portfolio/update_portfolio_usecase"

module CodingInterview
  module Presentation
    PortfolioItemDto = Struct.new(:symbol, :rate)
    GetOptimalPortfolioResponse = Struct.new(:portfolios)
    UpdateOptimalPortfolioRequest = Struct.new(:portfolios)

    class PortfolioController
      def initialize(get_latest_portfolio_usecase, update_portfolio_usecase)
        @get_latest_portfolio_usecase = get_latest_portfolio_usecase
        @update_portfolio_usecase = update_portfolio_usecase
      end

      def get_optimal_portfolio
        out = @get_latest_portfolio_usecase.run
        GetOptimalPortfolioResponse.new(
          out.items.map { |i| PortfolioItemDto.new(i.symbol.to_s, i.rate.to_s("F")) }
        )
      end

      def update_optimal_portfolio(req)
        items = req.portfolios.map do |dto|
          sym = Domain::StockSymbol.from_string(dto.symbol)
          raise BadRequestException.new("unknown symbol: #{dto.symbol}") if sym.nil?
          rate =
            begin
              BigDecimal(dto.rate)
            rescue ArgumentError, TypeError
              raise BadRequestException.new("invalid rate: #{dto.rate}")
            end
          Application::Usecase::Portfolio::UpdatePortfolioItemInput.new(sym, rate)
        end
        begin
          @update_portfolio_usecase.run(Application::Usecase::Portfolio::UpdatePortfolioUsecaseInput.new(items))
        rescue Application::Usecase::Portfolio::InvalidPortfolio => e
          raise BadRequestException.new(e.reason)
        end
      end
    end
  end
end
